model/riscv_vmem.sail

```diff
@@ -0,0 +1,857 @@
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits	   = 39,
+       vpn_size_bits	   = 9,
+
+       // PTE
+       levels		   = 3,
+       log_pte_size_bytes  = 3,	   // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+
+// PRIVATE
+let sv48_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits	   = 48,
+       vpn_size_bits	   = 9,
+
+       // PTE
+       levels		   = 4,
+       log_pte_size_bytes  = 3,	   // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+
+// For future; not currently used
+// PRIVATE
+/*
+let sv57_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits        = 57,
+       vpn_size_bits       = 9,
+
+       // PTE
+       levels              = 5,
+       log_pte_size_bytes  = 3,    // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+*/
+
+// PRIVATE
+val      undefined_SV_Params : unit -> SV_Params
+
+function undefined_SV_Params() = sv32_params
+
+// ****************************************************************
+// Fields of virtual addresses
+
+// PRIVATE: Extract full VPN field from VA
+function vpns_of_va(sv_params : SV_Params,
+		    va	      : bits(64)) -> bits(64) = {
+    let mask : bits(64) = (zero_extend(0b1) << sv_params.va_size_bits) - zero_extend(0b1);
+    (va & mask) >> PAGESIZE_BITS
+}
+
+// PRIVATE: Extract VPN[level] from  VA
+function vpn_j_of_va(sv_params : SV_Params,
+		     va	       : bits(64),
+		     level     : nat)	    -> bits(64) = {
+    let lsb  : nat	 = PAGESIZE_BITS + level * sv_params.vpn_size_bits;
+    let mask : bits(64) = (zero_extend(0b1) << sv_params.vpn_size_bits) - 1;
+    ((va >> lsb) & mask)
+}
+
+// PRIVATE: Extract offset within page from VA
+function offset_of_va(va : bits(64)) -> bits(64) = {
+    let mask : bits(64) = (zero_extend(0b1) << PAGESIZE_BITS) - 1;
+    va & mask
+}
+
+// Valid xlen-wide values containing virtual addrs must have upper
+// bits equal to the MSB of the virtual address.
+// Virtual address widths depend on the virtual memory mode.
+
+// PRIVATE
+function isValidVAddr(sv_params : SV_Params, vAddr : bits(64)) -> bool = {
+  let index = sv_params.va_size_bits - 1;
+  assert((0 <= index) & (index < 64));
+  let upper_bits = vAddr >> (sv_params.va_size_bits);
```

Comment:
Not sure if it's any better, but if you use a slice here instead of a shift you can drop the shift from the == ones() case.

Reply:
I tried the following using a slice, but it does not type-check:

    function isValidVAddr(sv_params : SV_Params, vAddr : bits(64)) -> bool = {
      let index = sv_params.va_size_bits - 1;
      assert((0 <= index) & (index < 64));
      let upper_bits = vAddr [63 .. sv_params.va_size_bits];
      let va_msb	 = vAddr[index];
      if va_msb == bitzero then
        upper_bits == zeros()
      else
        upper_bits == ones()
    }

I think the problem is that the slice does not have a size known statically to the type-checker; the size depends on the dynamic value sv_params.va_size_bits.  My technique avoids this by making everything 64 bits; the “dynamic” part goes into the shift-amount, which is ok, and does not affect type-checking.

Please let me know if my reasoning is ok, or if there is some other way to do this with a slice?

Reply:
I think the type system isn't quite clever enough to see that `sv_params.va_size_bits` is constant. It might work if you did something like:
```
function isValidVAddr(sv_params : SV_Params, vAddr : bits(64)) -> bool = {
  let index = sv_params.va_size_bits - 1;
  assert((0 <= index) & (index < 64));
  let upper_bits = vAddr [63 .. index + 1];
  let va_msb	 = vAddr[index];
  if va_msb == bitzero then
    upper_bits == zeros()
  else
    upper_bits == ones()
}
```
but not sure if that is clearer

Reply:
You could also pattern match on the struct, something like:
```
function isValidVAddr(struct { va_size_bits, _ }: SV_Params, vAddr : bits(64)) -> bool = {
  let index = va_size_bits - 1;
  assert((0 <= index) & (index < 64));
  let upper_bits = vAddr [63 .. va_size_bits];
  let va_msb	 = vAddr[index];
  if va_msb == bitzero then
    upper_bits == zeros()
  else
    upper_bits == ones()
}
```

Reply:
I don’t uderstand these last two suggestions.  va_size_bits can be 32, 39, 48 (and in future, 57), i.e., it’s still not a unique static value.  Is the type-checker smart enough to know (a) that it typechecks for each of those cases and (b) that zeros() and ones() needs to extended by a different amount for each of those cases?

Reply:
I also have a question about the struct pattern-match syntax in your example:

    struct { va_size_bits, _ }

I would have expected something like this:

    struct { va_size_bits = v, _ }

where 'va_size_bits' is the field name and 'v' is the local identifier bound to its value.  Is your syntax equivalent to:

    struct { va_size_bits = va_size_bits, _ }

where the field name is also used as the bound identifier?

Reply:
The type system keeps track of the numeric value of arguments at the level of individual variables, so if you bind `va_size_bits` as a variable once as in my last example with a struct pattern, it'll know it is either 32, 39, or 48, and importantly it'll know it's the same between the two usages, and it'll correctly infer the implicit arguments for zeros and ones so they end up the right size. The type of `upper_bits` is dependent on the value of `va_size_bits`, and it'll also know that `index` is whatever `va_size_bits` is minus 1, so the assert is actually not needed, and it can prove all the indexing is in bounds. I had a go on my machine and the following checks:

```
function isValidVAddr(struct { va_size_bits, _ }: SV_Params, vAddr : bits(64)) -> bool = {
  let index      = va_size_bits - 1;
  let upper_bits = vAddr[63 .. va_size_bits];
  let va_msb	 = vAddr[index];
  if va_msb == bitzero then
    upper_bits == zeros()
  else
    upper_bits == ones()
}
```

Since we only keep track of integer constraints for variables, in the code you wrote originally it doesn't know that `sv_params.va_size_bits` is the same for each of the two usages in the function, as that would require tracking constraints at a finer field-level granularity (which maybe we could do in the future, and Rust for example already does this type of field-level granularity checking for lifetimes).

Reply:
> where 'va_size_bits' is the field name and 'v' is the local identifier bound to its value. Is your syntax equivalent to:
> 
> ```
> struct { va_size_bits = va_size_bits, _ }
> ```
> 
> where the field name is also used as the bound identifier?

Yes exactly, a struct pattern `struct { foo, _ }` is shorthand for `struct { foo = foo }`.

Reply:
Ok, understood. Very cool. I’ve incorporated your suggested version.